File: models/a-p.rb
# -*- coding: UTF-8 -*-
require 'light_mongo'
require 'open-uri'
require 'nokogiri'

# LightMongo throws a RuntimeError if we don't have a default database set when
# we define a class that includes LightMongo::Document. Doing this in the test
# means that we have to require light_mongo explicitly, and then set the
# database inbetween that and requiring this file.  Eck.
begin
	LightMongo.database
rescue RuntimeError
	LightMongo.database = 'a-p_test'
end

module AP
	class Anime
		include LightMongo::Document
		
		index :slug
		
		attr_accessor :slug         # string (key)
		attr_accessor :type         # string
		attr_accessor :episodeCount # integer
		attr_accessor :startDate    # date
		attr_accessor :endDate      # date
		attr_accessor :description  # text
		
		def initialize(slug)
			@slug = slug
		end
		
		def update!
			doc = Nokogiri::HTML(open("http://www.anime-planet.com/anime/#{@slug}"))
			
			doc.at_css('.tabPanelLeft .type').content =~ /^(\w+) \((\d+\+?)/
			@type = $1
			@episodeCount = $2
			
			@description = doc.at_css('.entryContent .synopsis p').content
			
			doc.at_css('.tabPanelLeft .year').content =~ /(\d+|\?) - (\d+|\?)/
			# assign nil if the date is unknown
			@startDate = $1 == '?' ? nil : Date.strptime($1, '%Y')
			@endDate = $2 == '?' ? nil : Date.strptime($2, '%Y')
		end
	end
end

File: models/a-p.js
var APAnime = new Class({
	initialize: function(slug) {
		this.slug = slug;
		this.type =
		this.episodeCount =
		this.startDate =
		this.endDate =
		this.description = null;
	},
	
	update: function() {
		var anime = this;
		new Request.JSONP({
			url: 'http://fileserverness.ath.cx/jsonp.php?url=http://www.anime-planet.com/anime/'+this.slug,
			async: false, // this doesn't seem to actually work
			method: 'get',
			timeout: 3000,
			onSuccess: function(responseText){
				console.log('success!');
				var element, text, matches;

				element = new Element('div');
				element.set('html', responseText);

				text = element.getElement('.tabPanelLeft .type').get('text');
				matches = text.match(/^(\w+) \((\d+\+?)/);
				[, anime.type, anime.episodeCount] = matches;
				anime.episodeCount = parseInt(anime.episodeCount);
				
				anime.description = element.getElement('.entryContent .synopsis p').get('text');
				
				text = element.getElement('.tabPanelLeft .year').get('text');
				matches = text.match(/(\d+|\?) - (\d+|\?)/);
				[, anime.startDate, anime.endDate] = matches;
				anime.startDate = parseInt(anime.startDate);
				anime.endDate = parseInt(anime.endDate);
			},
			onFailure: function(xhr) {
				console.log('failure!');
				console.log(xhr);
			},
			onRequest: function() {
				console.log('request...');
			}
		}).send();
	}
});

